File: 015 3Sum.c
#include <stdio.h>
#include <stdlib.h>
/*
15 3Sum
https://oj.leetcode.com/problems/15 3Sum/
*/
int compInt(const void * a, const void * b)
{
  return ( *(int*)a - *(int*)b );
}
void sum3(int* arr, const int N){
    int k = 0,i,j;
    int lt,rt,md;
    
    qsort(arr,N,sizeof(int),compInt);
    while(k<N-2){
        lt = arr[k];
        i = k+1;
        j = N-1;
        while(i<j){
            md = arr[i];
            rt = arr[j];
            if(lt+md+rt==0){
                printf("(%d,%d,%d)\n",lt,md,rt);
                i++;
                while(i<j && arr[i]==md) i++;
                j--;
                while(i<j && arr[j]==rt) j--;
			}
            else if(lt+md+rt<0){
                i++;
                while(i<j && arr[i]==md) i++;
            }
            else{
                j--;
                while(i<j && arr[j]==rt) j--;
            }
        }
        k++;
        while(k<N-2 && arr[k]==lt) k++;
    }
    printf("\n\n");
}

int main(){
	//TODO: tests
    return 0; 
}

File: 015 3Sum.js
"use strict";
/*
3Sum
https://oj.leetcode.com/problems/15 3Sum/
*/
var threeSum = function(arr) {
    var pool = [];

    /*
    NOT:
    arr.sort();
    */
    arr.sort(function(x,y){
        return x - y;
    });
    
    var k = 0;
    while(k < arr.length - 2){
        var lt = arr[k];
        var i = k + 1;
        var j = arr.length - 1;
        while(i < j){
            var md = arr[i];
            var rt = arr[j];
            if(lt + md + rt === 0){
                pool.push([lt,md,rt]);
                
                i++;
                while(i<j && arr[i] === md) i++;
                j--;
                while(i<j && arr[j] === rt) j--;
            }
            else if(lt + md + rt < 0){
                i++;
                while(i<j && arr[i] === md) i++;
            }
            else{
                j--;
                while(i<j && arr[j] === rt) j--;
            }
        }
        k++;
        while(k < arr.length - 2 && arr[k] === lt) k++;
    }
    return pool;
};

var arr = [-1,-2,-3,4,5];
console.dir(JSON.stringify(threeSum(arr)));
